staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -161,29 +162,36 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
 		defer utilruntime.HandleCrash()
 		<-stopCh
 
-		lock.Lock()
-		defer lock.Unlock()
+		lock.RLock()
+		defer lock.RUnlock()
 		if client != nil {
 			client.Close()
 			clientErr = fmt.Errorf("server is shutting down")
 		}
 	}()
 
+	var requestGroup singleflight.Group
+	const checkETCDKey = "checkETCD"
 	return func() error {
 		// Given that client is closed on shutdown we hold the lock for
 		// the entire period of healthcheck call to ensure that client will
 		// not be closed during healthcheck.
 		// Given that healthchecks has a 2s timeout, worst case of blocking
 		// shutdown for additional 2s seems acceptable.
-		lock.Lock()
-		defer lock.Unlock()
+		lock.RLock()
+		defer lock.RUnlock()
 		if clientErr != nil {
 			return clientErr
 		}
-		ctx, cancel := context.WithTimeout(context.Background(), timeout)
-		defer cancel()
-		// See https://github.com/etcd-io/etcd/blob/c57f8b3af865d1b531b979889c602ba14377420e/etcdctl/ctlv3/command/ep_command.go#L118
-		_, err := client.Get(ctx, path.Join("/", c.Prefix, "health"))
+		// Coalesce concurrent requests to avoid excessive traffic from the
+		// healthz endpoint to etcd
+		_, err, _ := requestGroup.Do(checkETCDKey, func() (interface{}, error) {
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+			// See https://github.com/etcd-io/etcd/blob/c57f8b3af865d1b531b979889c602ba14377420e/etcdctl/ctlv3/command/ep_command.go#L118
+			_, err := client.Get(ctx, path.Join("/", c.Prefix, "health"))
+			return nil, err
+		})
```

Comment:
@liggitt this a more targeted fix, only the calls that need to get the request from etcd will be coalesce, we avoid any assumptions on the http middleware of the `healthz` endpoints

I saw in the other issue you linked that you were thinking on limit rating this check ... I can implement that option too, I don't see it will be hard, add a rate limit and serve from the cache the ones that overflow ...

Reply:
I thought more about this and I think it is better to use the rate limiter instead,